### PR TITLE
Improve Logging

### LINKED ISSUE #306: Better logging
The UEFI and bios loggers use only the framebuffer.

This can be an issue during headless runs (server with no display, unit tests)

I propose we refactor the logger to support multiple methods of logging output. (Serial being the first idea), with support in the kernel configuration to enable/disable sources. Or a text config file (this is likely better, so it can be configured at runtime)

This would also allow us to separate the logging and framebuffer init, as well be able to add/remove loggers at will. Rather than using `writeln` for early logging, we will be able to initialize the logger before anything else, and configure it with the optional `config.txt` or a default config that matches the current behavior.

### PR DESCRIPTION
**Pull request by asensio-project**
Hello,

This PR improves the actual logging of the bootloader.

Tasks:

- [x] Separate the logger from the framebuffer
- [x] Add support for Serial Being logger
- [x] Add configurations for enabling or disabling these methods of logging

Closes #306

Thanks!

### CODE REVIEW

**Comment by jasoncouture** [comment]
Makes sense. I'll also block `config.txt` on this as well, once the kernel config is in place.
It may yet again be wise to wait on #302 to avoid conflicts before doing the kernel config. So, perhaps do what is currently in this PR separately, as it'll be a huge help.

File: api/src/config.rs [hunk 1]
```diff
@@ -31,6 +31,12 @@ pub struct BootloaderConfig {
     /// Configuration for changing the level of the filter of the messages that are shown in the
     /// screen when booting. The default is 'Trace'.
     pub log_level: LevelFilter,
+
+    /// Configuration for enable or disable the framebuffer logger.
+    pub frame_buffer_logger_status: LoggerStatus,
+
+    /// Configuration for enable or disable the serial being logger.
+    pub serial_logger_status: LoggerStatus,
```
Comment (phil-opp) [inline api/src/config.rs:39]: Minor wording suggestions:

```suggestion

    /// Whether the bootloader should print log messages to the framebuffer when booting.
    ///
    /// Enabled by default.
    pub frame_buffer_logger_status: LoggerStatus,

    /// Whether the bootloader should print log messages to the serial port when booting.
    ///
    /// Enabled by default.
    pub serial_logger_status: LoggerStatus,
```

File: common/src/logger.rs [hunk 2]
```diff
@@ -1,71 +1,45 @@
...
-    }
-    get(c).unwrap_or_else(|| get(BACKUP_CHAR).expect("Should get raster of backup char."))
-}
+/// A [`FrameBufferWriter`] instance protected by a spinlock.
+pub struct LockedLogger(
+    Spinlock<FrameBufferWriter>,
+    Spinlock<SerialPort>,
+    LoggerStatus,
+    LoggerStatus,
+);
```
Comment (phil-opp) [inline common/src/logger.rs:16]: I would prefer to make this a normal struct with named fields. Also, we could directy resolve the `LoggerStatus` and set the fields as an `Option`:

```suggestion
/// A logger instance protected by a spinlock.
pub struct LockedLogger {
    framebuffer: OptionSpinlock<FrameBufferWriter>>,
    serial: Option<Spinlock<SerialPort>>,
}
```

File: common/src/logger.rs [hunk 3]
```diff
@@ -1,71 +1,45 @@
...
+        info: FrameBufferInfo,
+        frame_buffer_logger_status: LoggerStatus,
+        serial_logger_status: LoggerStatus,
+    ) -> Self {
+        LockedLogger(
+            Spinlock::new(FrameBufferWriter::new(framebuffer, info)),
+            Spinlock::new(SerialPort::new()),
+            frame_buffer_logger_status,
+            serial_logger_status,
+        )
```
Comment (phil-opp) [inline common/src/logger.rs:31]: As mentioned above:

```suggestion
    pub fn new(
        framebuffer: &'static mut [u8],
        info: FrameBufferInfo,
        frame_buffer_logger_status: LoggerStatus,
        serial_logger_status: LoggerStatus,
    ) -> Self {
        let framebuffer = match frame_buffer_logger_status {
            LoggerStatus::Enable => Some(Spinlock::new(FrameBufferWriter::new(framebuffer, info))),
            LoggerStatus::Disable => None,
        };
        let serial = match serial_logger_status {
            LoggerStatus::Enable => Some(SerialPort::new()),
            LoggerStatus::Disable => None,
        };
        LockedLogger {
            framebuffer,
            serial,
        }
```

File: common/src/logger.rs [hunk 4]
```diff
@@ -1,71 +1,45 @@
...
     /// Force-unlocks the logger to prevent a deadlock.
     ///
     /// ## Safety
     /// This method is not memory safe and should be only used when absolutely necessary.
     pub unsafe fn force_unlock(&self) {
-        unsafe { self.0.force_unlock() };
+        unsafe {
+            self.0.force_unlock();
+            self.1.force_unlock();
+        };
```
Comment (phil-opp) [inline common/src/logger.rs:42]: With the above changes, we could use `if let` here:

```suggestion
        if let Some(framebuffer) = &mut self.framebuffer {
            unsafe { framebuffer.force_unlock() };
        }
        if let Some(serial) = &mut self.serial {
            unsafe { serial.force_unlock() };
        }
```

File: common/src/logger.rs [hunk 5]
```diff
@@ -75,119 +49,15 @@ impl log::Log for LockedLogger {
...
-                }
-                let new_ypos =
-                    self.y_pos + font_constants::CHAR_RASTER_HEIGHT.val() + BORDER_PADDING;
-                if new_ypos >= self.height() {
-                    self.clear();
-                }
-                self.write_rendered_char(get_char_raster(c));
-            }
+        if self.2 == LoggerStatus::Enable {
+            let mut logger = self.0.lock();
```
Comment (phil-opp) [inline common/src/logger.rs:53]:
```suggestion
        if let Some(framebuffer) = &mut self.framebuffer {
            let mut logger = framebuffer.lock();
```

File: common/src/logger.rs [hunk 6]
```diff
@@ -75,119 +49,15 @@ impl log::Log for LockedLogger {
...
-
-    /// Prints a rendered char into the framebuffer.
-    /// Updates `self.x_pos`.
-    fn write_rendered_char(&mut self, rendered_char: RasterizedChar) {
-        for (y, row) in rendered_char.raster().iter().enumerate() {
-            for (x, byte) in row.iter().enumerate() {
-                self.write_pixel(self.x_pos + x, self.y_pos + y, *byte);
-            }
+        if self.3 == LoggerStatus::Enable {
+            let mut serial = self.1.lock();
```
Comment (phil-opp) [inline common/src/logger.rs:57]:
```suggestion
        if let Some(serial) = &mut self.serial {
            let mut serial = serial.lock();
```

**Review by phil-opp — COMMENTED** [review]
Looks very good! I only left some style comments.

**Comment by asensio-project** [comment]
I can't apply the changes proposed by @jasoncouture in the linked issue because it's very difficult to refactor BIOS stages due incompatibility problems of the used crates of the logger. So, I think that this PR is ready to merge.

**Review by phil-opp — APPROVED** [review]

**Comment by phil-opp** [comment]
Thanks!

**Pull request merged**